### PR TITLE
performance improvements to the position functions and tree lookup

### PR DESCRIPTION
**Pull request by AMarcedone**
@maxtaco I tried implementing my solution and here are the results.
We can also try with storing the path length with the key if you think it is worth it.

Also cc @heronhaye 

Here are the profiling results
[memprofile.pdf](https://github.com/keybase/client/files/3596941/memprofile.pdf)
[profile.pdf](https://github.com/keybase/client/files/3596942/profile.pdf)

### CODE REVIEW

File: go/merkletree2/tree.go [hunk 1]
```diff
@@ -18,17 +18,31 @@ type Tree struct {
 
 	cfg Config
 	eng StorageEngine
+
+	// step is an optimization parameter for GetKeyValuePairWithProof that
+	// controls how many path positions at a time the tree requests from the
+	// storage engine. Lower values result in more storage engine request, but
```
Comment (maxtaco) [inline go/merkletree2/tree.go]: request -> requests

File: go/merkletree2/tree.go [hunk 2]
```diff
@@ -18,17 +18,31 @@ type Tree struct {
...
+	// less of the (somewhat expensive) bit fiddling operations.  Values higher
+	// than 63 are not recommended as the bit operations (in the best case)
+	// cannot be done using a signle 64 bit word and become more expensive. This
```
Comment (maxtaco) [inline go/merkletree2/tree.go]: signle -> single

File: go/merkletree2/tree.go [hunk 3]
```diff
@@ -336,15 +350,31 @@ func (t *Tree) GetKeyValuePairWithProof(ctx context.Context, tr Transaction, s S
...
-	if err != nil {
-		return KeyValuePair{}, MerkleInclusionProof{}, err
-	}
-	// Assumes nodes are returned in the same order in which they are being requested
-	siblingPosHashPairs, err := t.eng.LookupNodes(ctx, tr, s, siblingPositions)
-	if err != nil {
-		return KeyValuePair{}, MerkleInclusionProof{}, err
+	var siblingPosHashPairs []PositionHashPair
+	needMore := true
+	for curr := 1; needMore && curr <= t.cfg.keysByteLength*8/int(t.cfg.bitsPerIndex); curr += t.step + 1 {
```
Comment (maxtaco) [inline go/merkletree2/tree.go]: can you make a function on `cfg` to do this division? would clear up the code

File: go/merkletree2/position.go [hunk 4]
```diff
@@ -103,25 +127,47 @@ func (t *Config) getDeepestPositionForKey(k Key) (*Position, error) {
...
+func (t *Config) getDeepestPositionAtLevelAndSiblingsOnPathToKey(k Key, lastLevel int, firstLevel int) (sibs []Position) {
+
+	maxLevel := t.keysByteLength * 8 / int(t.bitsPerIndex)
+	if lastLevel > maxLevel {
+		lastLevel = maxLevel
+	}
+
+	// first, shrink the key for efficiency
+	var bytesNecessary int
+	if lastLevel*int(t.bitsPerIndex)%8 == 0 {
```
Comment (maxtaco) [inline go/merkletree2/position.go]:
```go
bytesNecessary = lastLevel * int(t.bitsPerIndex) 
if lastLevel*int(t.bitsPerIndex)%8 != 0 {
    bytesNecessary++
}
```

Comment (AMarcedone) [reply go/merkletree2/position.go]: Is this for readability? or are there other reasons for the change? Also the first line should be 
` bytesNecessary = lastLevel * int(t.bitsPerIndex) / 8`  right?

Comment (maxtaco) [reply go/merkletree2/position.go]: readability

**Comment by AMarcedone** [comment]
Thanks for the quick review :)

**Pull request merged**